added startAppend() that appends new providers to old ones in stacked traces

techorix opened the pull request:
This is (roughly) a required feature for us at Notebooksbilliger.de and so fyi this pull request.

whiskeysierra commented:
First of all, thanks for contributing! Before we go into details, can you describe your use case a bit? What's the problem you're trying to solve?

techorix commented:
Yes, when using the stacked tracer we dont want to consider the previous tracer message provider every time we add a new tracer to the stack, so we just append the new message provider when starting a new tracer. 

This is only for easier handling of the tracer. 

The delimiter between tracer values could also be argument of the append method. 

If you want this feature, I can improve the pull request according to your suggestions.

whiskeysierra commented:
I was looking for a more high-level requirement. Under which circumstances do you need two (or more) nested traces? How is the current version of the library blocking you from doing that?

JonasJurczok commented on tracer-core/src/main/java/org/zalando/tracer/StackedTracer.java:
```diff
@@ -51,6 +51,30 @@ public void start(final Function<String, String> provider) {
+            queue.add(current);
+
+            runIf(listeners::onStop, name, previous);
+            runIf(listeners::onStart, name, current);
```
The problem is actually in this line.

JonasJurczok commented:
We talked about it and there was a misunderstanding :)

The issue is actually in the behaviour of the StackedTracer vs a (non existant) SuspendedTracer. Basically what we talked about a couple of weeks ago.

@techorix will update the PR with a new proposal

whiskeysierra commented:
> will update the PR with a new proposal

Before we do that, let's agree on a common understanding. In written form, this time :yum:

JonasJurczok commented:
:D okay.

The problem that we have is that we have a lot of nested traces in the system. This is currently done with suffixes. This is really horrible in regards to search.

What we want to do is use the Tracer lib to abstract the whole handling away from us but still have the stacking of traces available.

The ultimate goal is to enable us to follow subprocesses in the logs (until we have cleaned the system(tm)).

What we need to accomplish this is a change in behavior of the stacked tracer.

Lets assume an outer trace A and an inner trace B

Current behavior | Intended behavior
-------------------------------- | -------------
Start A | Start A
Stop A | Start B
Start B | Stop B
Stop B | Stop A
Start A |
Stop A |

Then the StackedMDCListener would do the right thing to our logs and we could use the logs to trace the sub processes in our system.

whiskeysierra commented:
> The problem that we have is that we have a lot of nested traces in the system.

So you're in fact transporting nested traces across service boundaries? Is there limit? Because the way I see this it won't scale beyond a certain point if we just append stuff. Clients will no longer be able to rely on a certain size of the trace identifiers. It all goes downwards from there.

JonasJurczok commented:
There is no hardcoded limit somewhere but it never goes beyond 4-5 from what I've seen so far.

whiskeysierra commented:
> it never goes beyond 4-5

Quite the risk if it's conceptually unlimited. Even 4-5 is already a size increase of 400-500% compared to single traces.

What do you use nested traces for?

JonasJurczok commented:
The problem is that there is a wild mixture between service boundaries and processes.

Basically what we want to do with the nested traces is being able to partition the logs by process and not just by class (even service is not possible at the moment).

JonasJurczok commented:
yes

techorix commented:
I updated the PR and added a StackedAppendTracer that does the same as the StackedTracer; except, it does not `runIf(listeners::onStop, name, previous);` in the start method. This does the same as the previous commit but employs the already built-in functionality of the stacked MDC listener.

JonasJurczok commented:
> I updated the PR and added a StackedAppendTracer that does the same as the StackedTracer; except, it does not runIf(listeners::onStop, name, previous); in the start method. This does the same as the previous commit but employs the already built-in functionality of the stacked MDC listener.

What about the stop method?
Also I think we where not finished with the discussion about what we actually want ;)

techorix commented:
Yes, we could also remove `runIf(listeners::onStart, name, current);` from the stop method. I could not verify an effect on functionality but it looks more like "stacking" when you look at the `StackedAppendTracerTest`.

Also sorry for just updating the PR. Did not read the conversion beforehand.  Though, StackedAppendTracerTest is probably the test class for modelling desired behaviour.

techorix commented:
from my understanding we want stacked tracers to transparently append new trace values to existing ones so that we can see the whole stack of traces in the Logs the MDC Listener writes to; vice versa, we want stack pops to remove trace values subsequently from newest to oldest

as the new trace has to be started somewhere (in a class) it should remain active until it is stopped (in the same or another class) carrying over the trace value (as specified above) in httprequests etc. until stopped

though I think tracers always have to be stopped in the same class they were started. As, otherwise, we would have to be able to verify which trace on which stack belongs to which call, right?

JonasJurczok commented:
so... @whiskeysierra whats your opinion on this? Is this something that would be acceptable for you?

Because then we could start discussing the implementation details :)

whiskeysierra commented:
> Is this something that would be acceptable for you?

The way *stacking* (what should've really be named *suspending*) works right now is purely *local*, i.e. it doesn't cross service boundaries. It was essentially just an easy solution to a minor issue we had because we're lazy developers: Splitting up batch processing into smaller chunks, while each chunks get's its own trace. Only the inner most trace is actually *active*, the outer ones are suspended.

That was the reason why I was (and still am) opposed to the idea to add it to the log (MDC). Your proposal definitely violates the idea of containing those *tricks* inside a service even further. I think you'd be better off with a solution like Zipkin/OpenTracing, check the [alternatives](https://github.com/zalando/tracer#alternatives). It nicely solves the issue of scalability because every *process* only needs to remember its *parent* and the *root*. With those two, limited information you can build a complete tree.

techorix commented:
In your case it would not make much difference if the batch chunks would have an optional common identifier but this might rather help identifying which chunk belongs to which batch. Wouldn't it?

Also, it is fine if only the top of the stack is active in our case and even in your case the trace does not need to remain completely local, right? I mean you could do a web operation after stacking and transmitted the newly started trace with the tracer. Couldn't you? I mean the `TracerHttpRequestInterceptor` doesnt care which implementation of the interface is transmitted?

JonasJurczok commented:
> Also, it is fine if only the top of the stack is active in our case and even in your case the trace does not need to remain completely local, right? I mean you could do a web operation after stacking and transmitted the newly started trace with the tracer. Couldn't you? I mean the TracerHttpRequestInterceptor doesnt care which implementation of the interface is transmitted?
 Hide all checks

No it makes a difference. Because with the way Tracer is implemented you can only pass one. For tracing all the existing processes and subprocesses you would have to pass 1-n traces at the same time.

techorix commented:
Okay, so the top trace value would have to contain our complete stack in a similar way to my first Pull Request or n-1 traces have to be transmitted. Guess I understand the problem now.

techorix commented:
@whiskeysierra I have problems verifying that the tracer transmits the generated Uuid between different services that use the tracer. is this also what you mean by "purely local"?

techorix commented:
@whiskeysierra I want to use the tracer to trace communication between two REST services running on Spring Boot Undertow. Can you verify that the tracer value (Uuid) is, or can be, transmitted between these services. Or do you mean by "does not cross service boundaries" that network distributed trace values are not within the scope of this software? Or how could this be achieved?

whiskeysierra commented:
> Can you verify that the tracer value (Uuid) is, or can be, transmitted between these services.

That's the main purpose of the project :stuck_out_tongue_winking_eye: 

> Or do you mean by "does not cross service boundaries" that network distributed trace values are not within the scope of this software?

That was meant for nested/stacked traces where we only transport the inner most trace to the next service.

> Or how could this be achieved?

Use the interceptor if you are using the Apache HTTP client: https://github.com/zalando/tracer#apache-http-client

techorix commented:
Okay, thanks for assuring. Ill see to making it work. Cool stuff :)

techorix commented:
we decided to use the tracer differently. maybe we will contribute new pull requests in the future.

best wishes :+1:
